notes/12-guarding-routes.md

```diff
@@ -0,0 +1,195 @@
+# Guarding Routes
+
+In this exercise, we will be adding redirection in the UI flow based on the availability of an authenticated session.
+Specifically, users should be able to access the `login` route, only when an authenticated session _does not exist_. And similarly, users should be able to access the `teams` route only when an authenticated _session exists_.
+
+So let's get started.
+
+## Changes Needed in Javascript
+
+First let's implement the auth service that we will be using to do session validation(or invalidation). This auth service is defined at `app/services/auth.js`.
+
+Import the action decorator at the top of the file:
+
+```js
+import { action } from '@ember/object';
+```
+
+Then add a property, `isAuthenticated` that will be used to store if the session is in an authenticated state, and an action method, `logout`, for invalidating the authenticated session and signing off the user.
+
+```js
+  get isAuthenticated() {
+    return !!this.currentUserId;
+  }
+
+  @action
+  logout() {
+    window.localStorage.removeItem(AUTH_KEY);
+    this.router.transitionTo('login');
+  }
+```
+
+In the login route, present at `app/routes/login.js`, add a `beforeModel` hook to perform session validations and handle the UI flow accordingly.
+
+In the login route, inject the imported `auth` service and add the `beforeModel` hook to perform session validation.
+
+```js
+   /**
+   * @type {AuthService}
+   */
+  @service auth;
+
+  async beforeModel(transition) {
+    await super.beforeModel(transition);
+    if (this.auth.isAuthenticated) {
+      this.transitionTo('teams');
+    }
+  }
+```
+
+And, in the same route file(`login` route), import the `auth` service, since its not already available(injected).
```

Comment:
Let's instruct readers to import first, then implement (so they get the nice autocomplete as they type).

We should also point out that `AuthService` is used in the JSDoc comment, to improve the editing experience (i.e., `this.auth.*` has a nice autocomplete).